docs: Update ECS docs for IAM auth method support (#13222)
hashicorp/consul: website/content/docs/ecs/terraform/secure-configuration.mdx

-> **NOTE:** This page assumes that you have already configured your Consul server with the above features.

## Auth Method

Consul on ECS uses the [AWS IAM Auth Method](/docs/security/acl/auth-methods/aws-iam) to enable
tasks to automatically obtain Consul ACL tokens during startup. Refer to the
[Architecture](/docs/ecs/architecture#aws-iam-auth-method) page for an overview of the auth method
with Consul on ECS.

With the Terraform modules for Consul on ECS, the auth method is supported by default when ACLs are enabled. The ACL controller sets up the auth method on the Consul servers. The `mesh-task` module configures the ECS task definition to be compatible with the auth method.

> A unique task IAM role is required for each ECS task family. Since the task family represents one Consul
> service, and since the task IAM role must encode the Consul service name, task IAM roles must not be shared
> by different task families.

Reviewer: ```suggestion
A unique task IAM role is required for each ECS task family. Task IAM roles must not be shared by different task families. This is because the task family represents only one Consul service and the task IAM role must encode the Consul service name. 
```
Better? Just a nit, but we should use "since" when describing a relationship with time and "because" when describing relationships pertaining to cause/effect.

-> **NOTE:** This page assumes that you have already configured your Consul server with the above features.

## Auth Method

Consul on ECS uses the [AWS IAM Auth Method](/docs/security/acl/auth-methods/aws-iam) to enable ECS
tasks to automatically obtain Consul ACL tokens during startup. See the [Architecture](/docs/ecs/architecture)
page for an overview of task startup, task shutdown, ACL tokens, and the AWS IAM auth method for Consul on ECS.

### Auth Method Configuration

When using the Terraform modules for Consul on ECS, the auth method is automatically configured
by the ACL controller, and ECS tasks will be configured by the `mesh-task` module to be able
to log in to the auth method.

> The `mesh-task` module configures the task IAM role as follows:

Reviewer 2: Do they need to ensure each service task family has a different task iam role?

Reviewer 2: Should we talk about how the mesh task module accepts a `task_role`  variable. If not passed, then we will automatically create valid roles.

If passed, then we will add tags to those roles.

* One instance of the AWS IAM auth method is configured for client tokens:
  * A Consul role is created with `node:write` permissions for client tokens used by Consul clients.
  * A binding rule is configured that attaches the Consul role to each token created during a successful login to this auth method instance.
* One instance of the AWS IAM auth method is configured for service tokens:
  * A binding rule attaches a [service identity](/docs/security/acl#service-identities) to each token created during a successful login to this auth method instance. The service name for this service identity is taken from the tag, `consul.hashicorp.com.service-name`, on the IAM role.
  * <EnterpriseAlert inline /> A namespace binding rule creates the service token in the namespace specified in the tag, <code>consul.hashicorp.com.namespace</code>, on the IAM role.
* Both instances of the auth method only permit IAM roles with an ARN matching this pattern, `arn:aws:iam::<ACCOUNT>:role/consul-ecs/*`. This allows IAM roles at the path `/consul-ecs/` to log in. The role path can be changed by setting the `iam_role_path` input variable for the `mesh-task` and `acl-controller` modules.

> 

Reviewer 2: This description seems to make more sense in the architecture section (and I think a lot of the content is already there?). For this page I think we should focus on what the user needs to do. So the last part about the IAM role needing to have a specific arn is important.

Maybe we can split this between "bringing your own task role" and "using the default task role"?

| `source`                            | string   | The location of the `acl-controller` module source.                                                                                                                                                                                        |
| `version`                           | string   | The version of the `acl-controller` module.                                                                                                                                                                                                |
| `consul_bootstrap_token_secret_arn` | string   | The Secrets Manager secret containing an ACL token for the ACL controller. This must contain `acl:write` and `operator:write` permissions.                                                                                                 |
| `consul_server_http_addr`           | string   | The HTTP(S) address of the Consul servers.                                                                                                                                                                                                 |

> | `consul_server_ca_cert_arn`         | string   | (optional) The Secrets Manager secret containing the CA cert for HTTPS communication with Consul servers. Required if the server's certificate is self-signed or signed by an internal CA. This is not required for Consul servers in HCP. |

Reviewer: ```suggestion
| `consul_server_ca_cert_arn`         | string   | (Optional) The Secrets Manager secret containing the CA cert for HTTPS communication with Consul servers. Required if the server's certificate is self-signed or signed by an internal CA. This is not required for Consul servers in HCP. |
```
This is the only field described as optional. Are the others required?

Author: Yep. The other inputs are required. I updated the sentence prior to this table to indicate that, but still call this one out as optional since it important but not always required. Let me know what you think.

| `gossip_key_secret_arn`     | string   | The Secrets Manager secret containing Consul's gossip encryption key.                                                                                                                                                                      |
| `acls`                      | boolean  | If true, ACLs are enabled.                                                                                                                                                                                                                 |
| `consul_http_addr`          | string   | The Consul server address. Required when `acls = true` in order to log in to Consul's AWS IAM auth method to obtain ACL tokens.                                                                                                            |
| `consul_https_ca_cert_arn`  | string   | (optional) The Secrets Manager secret containing the CA cert for HTTPS communication with Consul servers. Required if the server's certificate is self-signed or signed by an internal CA. This is not required for Consul servers in HCP. |

> Now you can deploy your services!

Reviewer: ```suggestion
You can now deploy your services.
```
Don't us exclamation points unless it is required a command or snippet of code.

| `gossip_key_secret_arn`     | string   | The Secrets Manager secret containing Consul's gossip encryption key.                                                                                                                                                                      |
| `acls`                      | boolean  | If true, ACLs are enabled.                                                                                                                                                                                                                 |
| `consul_http_addr`          | string   | The Consul server address. Required when `acls = true` in order to log in to Consul's AWS IAM auth method to obtain ACL tokens.                                                                                                            |
| `consul_https_ca_cert_arn`  | string   | (optional) The Secrets Manager secret containing the CA cert for HTTPS communication with Consul servers. Required if the server's certificate is self-signed or signed by an internal CA. This is not required for Consul servers in HCP. |

> Now you can deploy your services! Follow the rest of the steps in the [Installation instructions](/docs/ecs/terraform/install#task-module)

Reviewer 2: Should we have a section on how to control which tasks can be part of the mesh? Maybe talk about the security posture? So for example: "anyone with the ability to create iam roles and ecs tasks can create a task that joins the mesh wth any name" (not sure if that's true, but you get the idea. Also this could be done in a subsequent PR if you think it's a good idea)

Reviewer 2: Looks great!
